File: backend/app/api/tools.py
"""
Herramientas ARES — API lista para conectar UI.

Montaje:
    from app.api.tools import router as tools_router
    app.include_router(tools_router)

Endpoints:
    GET  /tools
    POST /tools/decode
    POST /tools/hash
    GET  /tools/cheatsheets/{os}          os = linux | windows
    GET  /tools/cheatsheets/{os}/{id}
"""
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from agent.core.audit import audit
from app.services import tools_service as svc

router = APIRouter(prefix="/tools", tags=["tools"])


# ── Request bodies ──────────────────────────────────────────────────────

class DecodeRequest(BaseModel):
    """Hash Decoder — decodifica Base64 / Hex / URL."""
    value: str = Field(..., min_length=1, max_length=100_000, description="Texto a decodificar")
    encoding: Literal["auto", "base64", "hex", "url"] = Field(
        default="auto",
        description="Formato; auto prueba base64 → hex → url",
    )


class HashRequest(BaseModel):
    """Hash Generator — MD5 / SHA1 / SHA256 (/ SHA512)."""
    value: str = Field(..., max_length=1_000_000, description="Texto plano (UTF-8)")
    algorithms: list[Literal["md5", "sha1", "sha256", "sha512"]] | None = Field(
        default=None,
        description="Default: md5, sha1, sha256",
    )


# ── Routes ──────────────────────────────────────────────────────────────

@router.get("")
def list_tools():
    """Catálogo de herramientas (para tabs / grid de la sección Herramientas)."""
    return svc.list_tools()


@router.post("/decode")
def decode_tool(body: DecodeRequest):
    """
    Hash Decoder.

    Ejemplo:
      POST /tools/decode
      { "value": "SGVsbG8=", "encoding": "base64" }
      → { "ok": true, "encoding": "base64", "output": "Hello", ... }
    """
    result = svc.decode_value(body.value, body.encoding)
    if result.get("ok"):
        audit.info(
            "tool_decode",
            "api",
            f"Decode {result.get('encoding')}: {len(body.value)} chars → {result.get('bytes_length')} bytes",
            {"encoding": result.get("encoding"), "ok": True},
        )
    else:
        audit.warn(
            "tool_decode",
            "api",
            f"Decode falló: {result.get('error')}",
            {"encoding": body.encoding, "ok": False},
        )
    return result


@router.post("/hash")
def hash_tool(body: HashRequest):
    """
    Hash Generator.

    Ejemplo:
      POST /tools/hash
      { "value": "Hello" }
      → { "ok": true, "hashes": { "md5": "...", "sha1": "...", "sha256": "..." } }
    """
    result = svc.generate_hashes(body.value, body.algorithms)
    if result.get("ok"):
        audit.info(
            "tool_hash",
            "api",
            f"Hash generado ({result.get('input_bytes')} bytes)",
            {"algorithms": list((result.get("hashes") or {}).keys()), "ok": True},
        )
    else:
        audit.warn(
            "tool_hash",
            "api",
            f"Hash falló: {result.get('error')}",
            {"ok": False},
        )
    return result


@router.get("/cheatsheets/{os_name}")
def cheatsheet_list(
    os_name: Literal["linux", "windows"],
    q: str | None = Query(default=None, max_length=120, description="Búsqueda (ej: systemctl)"),
    category: str | None = Query(default=None, max_length=60),
    limit: int = Query(default=50, ge=1, le=200),
):
    """
    Linux Cheat Sheet / Windows Commands — listado y búsqueda.

    Ejemplo:
      GET /tools/cheatsheets/linux?q=systemctl
    """
    result = svc.list_cheatsheet(os_name, q=q, category=category, limit=limit)
    if not result.get("ok"):
        raise HTTPException(status_code=400, detail=result.get("error") or "Error")
    return result


@router.get("/cheatsheets/{os_name}/{entry_id}")
def cheatsheet_detail(os_name: Literal["linux", "windows"], entry_id: str):
    """
    Detalle de un comando (explicación completa + ejemplos).

    Ejemplo:
      GET /tools/cheatsheets/linux/systemctl
    """
    result = svc.get_cheatsheet_entry(os_name, entry_id)
    if not result.get("ok"):
        raise HTTPException(status_code=404, detail=result.get("error") or "No encontrado")
    return result

File: frontend/src/lib/tools.js
/**
 * Cliente HTTP de Herramientas ARES (sin UI).
 * Conecta la sección "Herramientas" del frontend a /tools/*.
 *
 * Uso:
 *   import { listTools, decodeValue, generateHashes, searchLinux, searchWindows } from "../lib/tools"
 */
import { API_URL } from "./constants"

async function _json(res) {
  if (!res.ok) {
    let detail = `HTTP ${res.status}`
    try {
      const body = await res.json()
      detail = body.detail || body.error || detail
    } catch {}
    throw new Error(detail)
  }
  return res.json()
}

/** GET /tools — catálogo [{ id, title, status, endpoints, ... }] */
export async function listTools() {
  const res = await fetch(`${API_URL}/tools`)
  return _json(res)
}

/**
 * POST /tools/decode
 * @param {string} value - ej. "SGVsbG8="
 * @param {"auto"|"base64"|"hex"|"url"} [encoding="auto"]
 * @returns {{ ok, encoding, input, output, output_encoding, raw_hex, bytes_length, error? }}
 */
export async function decodeValue(value, encoding = "auto") {
  const res = await fetch(`${API_URL}/tools/decode`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ value, encoding }),
  })
  return _json(res)
}

/**
 * POST /tools/hash
 * @param {string} value - texto plano
 * @param {Array<"md5"|"sha1"|"sha256"|"sha512">} [algorithms]
 * @returns {{ ok, input, input_bytes, hashes: { md5, sha1, sha256 }, error? }}
 */
export async function generateHashes(value, algorithms) {
  const body = { value }
  if (algorithms?.length) body.algorithms = algorithms
  const res = await fetch(`${API_URL}/tools/hash`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
  return _json(res)
}

/**
 * GET /tools/cheatsheets/linux?q=
 * @param {string} [q]
 * @param {{ category?: string, limit?: number }} [opts]
 */
export async function searchLinux(q, opts = {}) {
  const params = new URLSearchParams()
  if (q) params.set("q", q)
  if (opts.category) params.set("category", opts.category)
  if (opts.limit) params.set("limit", String(opts.limit))
  const qs = params.toString()
  const res = await fetch(`${API_URL}/tools/cheatsheets/linux${qs ? `?${qs}` : ""}`)
  return _json(res)
}

/**
 * GET /tools/cheatsheets/windows?q=
 */
export async function searchWindows(q, opts = {}) {
  const params = new URLSearchParams()
  if (q) params.set("q", q)
  if (opts.category) params.set("category", opts.category)
  if (opts.limit) params.set("limit", String(opts.limit))
  const qs = params.toString()
  const res = await fetch(`${API_URL}/tools/cheatsheets/windows${qs ? `?${qs}` : ""}`)
  return _json(res)
}

/** GET /tools/cheatsheets/{os}/{entryId} */
export async function getCheatsheetEntry(os, entryId) {
  const res = await fetch(`${API_URL}/tools/cheatsheets/${os}/${encodeURIComponent(entryId)}`)
  return _json(res)
}

export const getLinuxEntry = (id) => getCheatsheetEntry("linux", id)
export const getWindowsEntry = (id) => getCheatsheetEntry("windows", id)
